Cannot parse closure in closure without space

A closure returning a closure without a space cannot be parsed by GCCRS.

I tried this code:

```rust
fn main() {
    let f = |_||x, y| x+y;
    assert_eq!(f(())(1, 2), 3);
}
```

:warning: The default rust formatter will change the code and introduce a space between the pipes, circumventing the bug (the code with the space compiles correctly).

I expected to see this happen: Compile correctly.

Instead, this happened:

```
./test.rs:2:15: error: expecting ‘|’ but ‘||’ found
    2 |     let f = |_||x, y| x+y;
      |               ^~
./test.rs:2:15: error: failed to parse expression in let statement
./test.rs:2:5: error: failed to parse statement or expression in block expression
    2 |     let f = |_||x, y| x+y;
      |     ^~~
./test.rs:4:1: error: unrecognised token ‘}’ for start of item
    4 | }
      | ^
./test.rs:4:1: error: failed to parse item in crate
```

### Meta

- c7728c8020334fecb47789163ba94326aab6208c

--- a/gcc/rust/parse/rust-parse-impl.h
+++ b/gcc/rust/parse/rust-parse-impl.h
@@ -14445,11 +14445,16 @@ Parser<ManagedTokenSource>::parse_closure_expr_pratt (const_TokenPtr tok,
 
 	    if (lexer.peek_token ()->get_id () != COMMA)
 	      {
+		if (lexer.peek_token ()->get_id () == OR)
+		  lexer.split_current_token (PIPE, PIPE);
 		// not an error but means param list is done
 		break;
 	      }
 	    // skip comma
 	    lexer.skip_token ();
+
+	    if (lexer.peek_token ()->get_id () == OR)
+	      lexer.split_current_token (PIPE, PIPE);
 
 	    t = lexer.peek_token ();
 	  }

--- a/gcc/testsuite/rust/compile/closure_in_closure.rs
+++ b/gcc/testsuite/rust/compile/closure_in_closure.rs
@@ -1,0 +1,8 @@
+// { dg-additional-options "-frust-compile-until=ast" }
+//
+// Do not reformat this test! The default rust format settings will insert a
+// space between closure parameter lists.
+fn main() {
+    let f = |_||x, y| x+y;
+    assert_eq!(f(())(1, 2), 3);
+}

